staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_width.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"math"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const (
+	// the minimum number of seats a list request occupies
+	minimumSeatsList uint = 3
+
+	// the maximum number of seats a list request occupies
+	maximumSeatsList uint = 10
+)
+
+func newListWidthEstimator(countFn objectCountGetterFunc) WidthEstimatorFunc {
+	estimator := &listWidthEstimator{
+		countFn: countFn,
+	}
+	return estimator.estimate
+}
+
+type listWidthEstimator struct {
+	countFn objectCountGetterFunc
+}
+
+func (e *listWidthEstimator) estimate(r *http.Request) Width {
+	requestInfo, ok := apirequest.RequestInfoFrom(r.Context())
+	if !ok {
+		return Width{Seats: minimumSeatsList}
+	}
+
+	query := r.URL.Query()
+	count := e.countFn.Get(key(requestInfo))
+
+	var fieldSelectorPresent bool
+	if len(query.Get("fieldSelector")) > 0 {
+		fieldSelectorPresent = true
+	}
```

Comment:
I'm actually afraid of introducing inconsistencies of this logic with the actual parsing - I would much rather prefer reusing the actual parsing logic. So something like
```
listOptions := metav1.ListOptions{}
if err := Convert_url_Values_To_v1_ListOptions(r.URL.Query(), &listOptions, nil); err != nil {
  // log error [the request should be rejected later anyway]
}

hasFieldSelector := len(listOptions.FieldSelector) > 0
hasLabelSelector := len(listOptions.LabelSelector) > 0
```